sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/JavaOrcColumnarBatchReader.java

```diff
@@ -0,0 +1,503 @@
+import org.apache.spark.sql.vectorized.ColumnarBatch;
+
+
+/**
+ * To support vectorization in WholeStageCodeGen, this reader returns ColumnarBatch.
+ * After creating, `initialize` and `setRequiredSchema` should be called sequentially.
+ */
+public class JavaOrcColumnarBatchReader extends RecordReader<Void, ColumnarBatch> {
+
+  /**
+   * ORC File Reader.
+   */
+  private Reader reader;
+
+  /**
+   * Vectorized Row Batch.
+   */
+  private VectorizedRowBatch batch;
+
+  /**
+   * Requested Column IDs.
+   */
+  private int[] requestedColIds;
+
+  /**
+   * Record reader from row batch.
+   */
+  private org.apache.orc.RecordReader recordReader;
+
+  /**
+   * Required Schema.
+   */
+  private StructType requiredSchema;
+
+  /**
+   * ColumnarBatch for vectorized execution by whole-stage codegen.
+   */
+  private ColumnarBatch columnarBatch;
+
+  /**
+   * Writable column vectors of ColumnarBatch.
+   */
+  private WritableColumnVector[] columnVectors;
+
+  /**
+   * The number of rows read and considered to be returned.
+   */
+  private long rowsReturned = 0L;
+
+  /**
+   * Total number of rows.
+   */
+  private long totalRowCount = 0L;
+
+  @Override
+  public Void getCurrentKey() throws IOException, InterruptedException {
+    return null;
+  }
+
+  @Override
+  public ColumnarBatch getCurrentValue() throws IOException, InterruptedException {
+    return columnarBatch;
+  }
+
+  @Override
+  public float getProgress() throws IOException, InterruptedException {
+    return (float) rowsReturned / totalRowCount;
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException {
+    return nextBatch();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (columnarBatch != null) {
+      columnarBatch.close();
+      columnarBatch = null;
+    }
+    if (recordReader != null) {
+      recordReader.close();
+      recordReader = null;
+    }
+  }
+
+  /**
+   * Initialize ORC file reader and batch record reader.
+   * Please note that `setRequiredSchema` is needed to be called after this.
+   */
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
+      throws IOException, InterruptedException {
+    FileSplit fileSplit = (FileSplit)inputSplit;
+    Configuration conf = taskAttemptContext.getConfiguration();
+    reader = OrcFile.createReader(
+      fileSplit.getPath(),
+      OrcFile.readerOptions(conf)
+        .maxLength(OrcConf.MAX_FILE_LENGTH.getLong(conf))
+        .filesystem(fileSplit.getPath().getFileSystem(conf)));
+
+    Reader.Options options =
+            OrcInputFormat.buildOptions(conf, reader, fileSplit.getStart(), fileSplit.getLength());
+    recordReader = reader.rows(options);
+    totalRowCount = reader.getNumberOfRows();
+  }
+
+  /**
+   * Set required schema and partition information.
+   * With this information, this creates ColumnarBatch with the full schema.
+   */
+  public void setRequiredSchema(
```

Comment:
how about we follow parquet and call it `initBatch`?

Reply:
Good idea. It's removed.